### PR TITLE
Add Speaker notes string.md

### PR DESCRIPTION
**Pull request by cchiw**
#274

### CODE REVIEW

File: src/std/string.md [hunk 1]
```diff
@@ -27,8 +27,14 @@ fn main() {
 
 <details>
 
+* `new` heap allocated buffer. `StringL::with capacity` is used when you know how much you want to push on
 * `len` returns the size of the `String` in bytes, not its length in characters.
 * `chars` returns an iterator over the actual characters.
-* `String` implements `Deref<Target = str>` which transparently gives it access to `str`'s methods.
+*  When people refer to strings they could either be talking about &str or String.
```
Comment (mgeisler) [inline src/std/string.md:33]:
```suggestion
*  When people refer to strings they could either be talking about `&str` or `String`. 
```

File: src/std/string.md [hunk 2]
```diff
@@ -27,8 +27,14 @@ fn main() {
...
+* `new` heap allocated buffer. `StringL::with capacity` is used when you know how much you want to push on
 * `len` returns the size of the `String` in bytes, not its length in characters.
 * `chars` returns an iterator over the actual characters.
-* `String` implements `Deref<Target = str>` which transparently gives it access to `str`'s methods.
+*  When people refer to strings they could either be talking about &str or String. 
+* Implementing Deref (trait), gives the compiler ability to take a value of any type, call the deref method, and know how to dereference it  
+    * `String` implements `Deref<Target = str>` which transparently gives it access to `str`'s methods.
+    * Write and compare `let s3 = s1.deref();` and  `let s3 = &*s1`;.
+* String isimplemented as a wrapper around a vector of bytes, many of the operations you see supported on vectors are also supported on String, but with some extra guarantees.
+* Show the danger of indexing Strings by (1) adding a unicode character to an above string (Ex. `Hβello`) and slicing it at various locations and (2) unwraping the characters `.chars.nth(i).unwrap()` where `i` is in-bound and out-of-bounds given the number of chars.
```
Comment (mgeisler) [inline src/std/string.md:38]: I think this is outdated now: we already have a Unicode flag (🇨🇭) above.

File: src/std/string.md [hunk 3]
```diff
@@ -27,8 +27,14 @@ fn main() {
...
 
+* `new` heap allocated buffer. `StringL::with capacity` is used when you know how much you want to push on
 * `len` returns the size of the `String` in bytes, not its length in characters.
 * `chars` returns an iterator over the actual characters.
-* `String` implements `Deref<Target = str>` which transparently gives it access to `str`'s methods.
+*  When people refer to strings they could either be talking about &str or String. 
+* Implementing Deref (trait), gives the compiler ability to take a value of any type, call the deref method, and know how to dereference it  
+    * `String` implements `Deref<Target = str>` which transparently gives it access to `str`'s methods.
+    * Write and compare `let s3 = s1.deref();` and  `let s3 = &*s1`;.
+* String isimplemented as a wrapper around a vector of bytes, many of the operations you see supported on vectors are also supported on String, but with some extra guarantees.
```
Comment (mgeisler) [inline src/std/string.md:37]:
```suggestion
* `String` is implemented as a wrapper around a vector of bytes, many of the operations you see supported on vectors are also supported on `String`, but with some extra guarantees.
```

File: src/std/string.md (hunk 2 above)
Comment (cchiw) [reply src/std/string.md]: I updated the text. Let me know what you think.

File: src/std/string.md [hunk 4]
```diff
@@ -27,7 +27,14 @@ fn main() {
 
 <details>
 
+* `new()` heap allocated buffer. `StringL::with capacity` is used when you know how much you want to push on
```
Comment (mgeisler) [inline src/std/string.md:30]:
```suggestion
* `String::new` returns a new empty string, use `String::with capacity` when you know how much data you want to push to the string.
```

File: src/std/string.md [hunk 5]
```diff
@@ -27,7 +27,14 @@ fn main() {
...
 * `len()` returns the size of the `String` in bytes, not its length in characters.
```
Comment (mgeisler) [inline src/std/string.md:31]:
```suggestion
* `String::len` returns the size of the `String` in bytes (which can be different from its length in characters).
```

File: src/std/string.md [hunk 6]
```diff
@@ -27,7 +27,14 @@ fn main() {
...
 * `chars()` returns an iterator over the actual characters.
+*  When people refer to strings they could either be talking about `&str` or `String`. 
+* Implementing Deref (trait), gives the compiler ability to take a value of any type, call the deref method, and know how to dereference it
```
Comment (mgeisler) [inline src/std/string.md:34]:
```suggestion
* When a type implements `Deref<Target = T>`, the compiler will let you transparently call methods from `T`.
```

File: src/std/string.md [hunk 7]
```diff
@@ -13,7 +13,7 @@ fn main() {
     s2.push('!');
     println!("s2: len = {}, capacity = {}", s2.len(), s2.capacity());
 
-    let s3 = String::from("🇨🇭");
+    let s3 = String::from("A Flag:🇨🇭");
```
Comment (mgeisler) [inline src/std/string.md:16]: We need to be able to reason about the string length at a glance in both bytes and characters. So we should not add extra text here:

```suggestion
    let s3 = String::from("🇨🇭");
```

File: src/std/string.md [hunk 8]
```diff
@@ -27,7 +27,14 @@ fn main() {
 
 <details>
 
+* `new()` heap allocated buffer. `StringL::with capacity` is used when you know how much you want to push on
 * `len()` returns the size of the `String` in bytes, not its length in characters.
 * `chars()` returns an iterator over the actual characters.
```
Comment (mgeisler) [inline src/std/string.md:32]:
```suggestion
* `String::chars` returns an iterator over the actual characters. Note that a `char` can be different from what a human will consider a "character" due to [grapheme clusters](https://docs.rs/unicode-segmentation/latest/unicode_segmentation/struct.Graphemes.html).
```

**Review by mgeisler — APPROVED** [review]

**Pull request merged**